Add drake visualization infrastructure

SeanCurtis-TRI opened the pull request:
Adds the infrastructure to connect geometry system with drake visualizer.
  1. Add output connections from `GeometrySystem` that can feed into ` DrakeVisualizer`.
  2. Add interface for dispatching LCM load message.
  3. Finish wiring up registration from system to state.
  4. Extend state to allow:
      1. Computation of frame position in world.
      2. Iterate through the frames in the system.

NOTE: This includes code that doesn't reflect the final state of functionality. THere are members of `GeometrySystem` and `GeometryState` that exist the way they do because they are waiting on the cache system.

The next PR will actually produce moving bodies that can be seen!

Breakdown of changes:
```
Category            added  modified  removed  
----------------------------------------------
code                506    10        2        
comments            133    7         1        
blank               94     0         0        
----------------------------------------------
TOTAL               733    17        3
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/6953)
<!-- Reviewable:end -->

SeanCurtis-TRI commented:
+@avalenzu for feature review, please.

---

Review status: 0 of 11 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-:-Ksz8Iur6IMd9IYXuGUB:b-fcohnv)*
<!-- Sent from Reviewable.io -->

avalenzu commented:
Checkpoint.

---

Reviewed 5 of 11 files at r1.
Review status: 5 of 11 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/geometry/geometry_state.h, line 125 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-KtMNYUP0W0MzfvqjO_x:-KtMNYUP0W0MzfvqjO_y:b-d177vy) ([raw file](https://github.com/robotlocomotion/drake/blob/b22ef77929959ba7f0660158d139191324c43c97/drake/geometry/geometry_state.h#L125)):*
> ```objc
>   }
> 
>   /** Reports the pose of the frame with the given id. */
> ```

This comment should mention that the returned pose is relative to the parent. Right now it's identical to the one for `get_pose_in_world()`

---

*[drake/geometry/geometry_state.h, line 373 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-KtMNG0-O7OaFAELlAoV:-KtMNG0-O7OaFAELlAoW:b2bbmvn) ([raw file](https://github.com/robotlocomotion/drake/blob/b22ef77929959ba7f0660158d139191324c43c97/drake/geometry/geometry_state.h#L373)):*
> ```objc
>     };
> 
>     explicit MapKeyIterator(std::unordered_map<K, V> map) : map_(map) {}
> ```

This copies the input, right? Is that what you want?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-:-KtMOOd-YEar4AiVZeXz:by96gdd)*
<!-- Sent from Reviewable.io -->

avalenzu commented:
Review pass done (finally - thanks for your patience!). PTAL.

---

Reviewed 6 of 11 files at r1.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[drake/geometry/geometry_system.cc, line 254 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-KtqwQucJ8eSB71Bx8Qd:-KtqwQud7ALR4ovu0jCk:bwdu6n6) ([raw file](https://github.com/robotlocomotion/drake/blob/b22ef77929959ba7f0660158d139191324c43c97/drake/geometry/geometry_system.cc#L254)):*
>           "Source " + to_string(source_id) + " has registered frames "
>               "but does not provide pose values on the input port.");

It'd be nice to get rid of the repetition of this string.

---

*[drake/geometry/geometry_system.h, line 98 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-Ktqr4ua6ggS38G5ERfD:-Ktqr4ub46hI34Ndqef9:bptxpac) ([raw file](https://github.com/robotlocomotion/drake/blob/b22ef77929959ba7f0660158d139191324c43c97/drake/geometry/geometry_system.h#L98)):*
> ```objc
> 
>  __lcm visualization port__: An abstract-valued port containing an instance of
>  PoseBundle. This is a convenience port designed to feed LCM update messages to
> ```

It might be better to replace "director" with "drake_visualizer," since we don't really expose the full director in our build targets.g

---

*[drake/geometry/geometry_visualization.cc, line 59 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-Ktr04hhT4X4aTJqcdhI:-Ktr04hhT4X4aTJqcdhJ:b-3a1onu) ([raw file](https://github.com/robotlocomotion/drake/blob/b22ef77929959ba7f0660158d139191324c43c97/drake/geometry/geometry_visualization.cc#L59)):*
> ```cpp
> 
>   void ImplementGeometry(const HalfSpace&) override {
>     // Currently representing a half space as a big box. This assumes that the
> ```

FYI: We should see if we can figure out a way for half-spaces at z = 0 to look better in director. The artifacts caused by the coincident grid and box surface are annoying.

---

*[drake/geometry/geometry_visualization.cc, line 70 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-KtrIwDScIswmRioTuPt:-KtrIwDScIswmRioTuPu:bk66mtu) ([raw file](https://github.com/robotlocomotion/drake/blob/b22ef77929959ba7f0660158d139191324c43c97/drake/geometry/geometry_visualization.cc#L70)):*
> ```cpp
> 
>     // The final pose of the box is the half-space's pose pre-multiplied by
>     // an offset sufficient o move the box down so it's top face lies on the
> ```

Typo "o" -> "to"

---

*[drake/geometry/geometry_visualization.h, line 14 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-KtrKFEZfHorRDcFTnUn:-KtrKFEZfHorRDcFTnUo:b-9v3pgl) ([raw file](https://github.com/robotlocomotion/drake/blob/b22ef77929959ba7f0660158d139191324c43c97/drake/geometry/geometry_visualization.h#L14)):*
> i.e., those entities registered on the system _prior_ to the allocation of the context

_Can_ entities be registered after the allocation of the context? I thought that wasn't allowed.

---

*[drake/geometry/test/geometry_system_test.cc, line 107 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-KtrKr1njz02C2DC3ul0:-KtrKr1njz02C2DC3ul1:b-1wnmec) ([raw file](https://github.com/robotlocomotion/drake/blob/b22ef77929959ba7f0660158d139191324c43c97/drake/geometry/test/geometry_system_test.cc#L107)):*
> ```cpp
>   // Ownership of context.
>   unique_ptr<Context<double>> context_;
>   // Direct access to a pre-cast geometry context-typed version of context_.
> ```

Can you add commas to clear up what the sequence of adjectives is here? Also the hyphenation of "geometry context-typed" looks odd to me.

---

*[drake/geometry/test/geometry_system_test.cc, line 445 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-KtrM1F3tLHwB5MO7r8N:-KtrM1F3tLHwB5MO7r8O:bgaqe8) ([raw file](https://github.com/robotlocomotion/drake/blob/b22ef77929959ba7f0660158d139191324c43c97/drake/geometry/test/geometry_system_test.cc#L445)):*
> ```cpp
> }
> 
> // Adversarial test case: Missing all port connection.
> ```

Typo: "connections"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-:-KtrM91bOfBpJ0q_Y-3G:b-xdjfch)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Done PTAL

---

Review status: 5 of 11 files reviewed at latest revision, 9 unresolved discussions.

---

*[drake/geometry/geometry_state.h, line 125 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-KtMNYUP0W0MzfvqjO_x:-KtxILnoLMrzG2GQqXeo:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/b22ef77929959ba7f0660158d139191324c43c97/drake/geometry/geometry_state.h#L125)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

This comment should mention that the returned pose is relative to the parent. Right now it's identical to the one for `get_pose_in_world()`
</blockquote></details>

Done

---

*[drake/geometry/geometry_state.h, line 373 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-KtMNG0-O7OaFAELlAoV:-KtxKibdF53EJDvZw2PF:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/b22ef77929959ba7f0660158d139191324c43c97/drake/geometry/geometry_state.h#L373)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

This copies the input, right? Is that what you want?
</blockquote></details>

Done

---

*[drake/geometry/geometry_system.cc, line 254 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-KtqwQucJ8eSB71Bx8Qd:-KtxLCOb5ZIKAJqqP7eL:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/b22ef77929959ba7f0660158d139191324c43c97/drake/geometry/geometry_system.cc#L254)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

>           "Source " + to_string(source_id) + " has registered frames "
>               "but does not provide pose values on the input port.");

It'd be nice to get rid of the repetition of this string.
</blockquote></details>

Done

---

*[drake/geometry/geometry_system.h, line 98 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-Ktqr4ua6ggS38G5ERfD:-KtxLK14dFp53hhiW48p:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/b22ef77929959ba7f0660158d139191324c43c97/drake/geometry/geometry_system.h#L98)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

It might be better to replace "director" with "drake_visualizer," since we don't really expose the full director in our build targets.g
</blockquote></details>

Done

---

*[drake/geometry/geometry_visualization.cc, line 59 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-Ktr04hhT4X4aTJqcdhI:-KtxLO8r4ThdcGl27AbJ:blt097o) ([raw file](https://github.com/robotlocomotion/drake/blob/b22ef77929959ba7f0660158d139191324c43c97/drake/geometry/geometry_visualization.cc#L59)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

FYI: We should see if we can figure out a way for half-spaces at z = 0 to look better in director. The artifacts caused by the coincident grid and box surface are annoying.
</blockquote></details>

OK I agree. Ultimately, the cause is z-buffer issues.  But what ya gonna do? It'd also be nice to not have to draw boxes.

---

*[drake/geometry/geometry_visualization.cc, line 70 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-KtrIwDScIswmRioTuPt:-KtxLYt2L84fRhhDGhaC:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/b22ef77929959ba7f0660158d139191324c43c97/drake/geometry/geometry_visualization.cc#L70)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

Typo "o" -> "to"
</blockquote></details>

Done

---

*[drake/geometry/geometry_visualization.h, line 14 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-KtrKFEZfHorRDcFTnUn:-KtxLk8L062JgcU9kuOh:b-q1jtsr) ([raw file](https://github.com/robotlocomotion/drake/blob/b22ef77929959ba7f0660158d139191324c43c97/drake/geometry/geometry_visualization.h#L14)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

> i.e., those entities registered on the system _prior_ to the allocation of the context

_Can_ entities be registered after the allocation of the context? I thought that wasn't allowed.
</blockquote></details>

Ultimately -- yes. When we get to the point where we can introduce/remove geometries. However, I think that detail does little good here. So, I've re-worded it.

---

*[drake/geometry/test/geometry_system_test.cc, line 107 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-KtrKr1njz02C2DC3ul0:-KtxMFYiuu7lAm8qDmRh:b-criy89) ([raw file](https://github.com/robotlocomotion/drake/blob/b22ef77929959ba7f0660158d139191324c43c97/drake/geometry/test/geometry_system_test.cc#L107)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

Can you add commas to clear up what the sequence of adjectives is here? Also the hyphenation of "geometry context-typed" looks odd to me.
</blockquote></details>

Also added an otherwise missing hyphen.

---

*[drake/geometry/test/geometry_system_test.cc, line 445 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-KtrM1F3tLHwB5MO7r8N:-KtxMKrkkB8Km15a_LE1:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/b22ef77929959ba7f0660158d139191324c43c97/drake/geometry/test/geometry_system_test.cc#L445)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

Typo: "connections"
</blockquote></details>

Done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-:-KtxMLqnLk7_Lkg-hV3U:b-q5tim0)*
<!-- Sent from Reviewable.io -->

avalenzu commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> +@soonho-tri for platform review.

---

Reviewed 6 of 6 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-:-Ku04BiKLx8FJTDsWxZi:b-sf1n7o)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Check point. PTAL.

---

Reviewed 4 of 11 files at r1, 4 of 6 files at r2.
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

*[drake/geometry/geometry_context.h, line 19 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-Ku26ZQSNeGKDq1RP9Iw:-Ku26ZQSNeGKDq1RP9Ix:b68p1t8) ([raw file](https://github.com/robotlocomotion/drake/blob/00cc27b59280badb1dba8c0a0e1429d804a75bb3/drake/geometry/geometry_context.h#L19)):*
> ```objc
>  No other values for T are currently supported. */
> template <typename T>
> class GeometryContext : public drake::systems::LeafContext<T> {
> ```

Can it be marked as `final`?

---

*[drake/geometry/geometry_context.h, line 19 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-Ku26bQgjTpcpqRYgCbU:-Ku26bQgjTpcpqRYgCbV:b-qqzx78) ([raw file](https://github.com/robotlocomotion/drake/blob/00cc27b59280badb1dba8c0a0e1429d804a75bb3/drake/geometry/geometry_context.h#L19)):*
> ```objc
>  No other values for T are currently supported. */
> template <typename T>
> class GeometryContext : public drake::systems::LeafContext<T> {
> ```

FYI, `drake::` is redundant since we're in the namespace.

---

*[drake/geometry/geometry_context.h, line 20 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-Ku26o3My0lZStzsPgqA:-Ku26o3My0lZStzsPgqB:b-svcqyt) ([raw file](https://github.com/robotlocomotion/drake/blob/00cc27b59280badb1dba8c0a0e1429d804a75bb3/drake/geometry/geometry_context.h#L20)):*
> ```objc
> template <typename T>
> class GeometryContext : public drake::systems::LeafContext<T> {
>  public:
> ```

Can we specify copy/move operations explicitly (i.e. using the macros defined in `drake_copyable.h`) as recommended in Google's C++ guide?

---

*[drake/geometry/geometry_context.h, line 32 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-Ku25wQfLVwASNsaDCwB:-Ku25wQfLVwASNsaDCwC:b-nugwpm) ([raw file](https://github.com/robotlocomotion/drake/blob/00cc27b59280badb1dba8c0a0e1429d804a75bb3/drake/geometry/geometry_context.h#L32)):*
> ```objc
>  private:
>   // The index of the geometry state abstract state.
>   int geometry_state_index_;
> ```

FYI, POD member fields need `{}`. 
BTW, this member field can be marked as `const`.

---

*[drake/geometry/geometry_state.h, line 111 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-Ku2Nd2iOFPv4lymMSVJ:-Ku2Nd2iOFPv4lymMSVK:b6av4pp) ([raw file](https://github.com/robotlocomotion/drake/blob/00cc27b59280badb1dba8c0a0e1429d804a75bb3/drake/geometry/geometry_state.h#L111)):*
> ```objc
> 
>   /** Reports the name of the frame. */
>   const std::string& get_frame_name(FrameId frame_id) const {
> ```

Returning `const T&` can be efficient but it can cause a problem if one uses a reference after the referenced frame was removed. Could you share your thought regarding this concern?

---

*[drake/geometry/geometry_state.h, line 355 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-Ku2SxcvNoBpLXyhaFpe:-Ku2SxcvNoBpLXyhaFpf:b-c1yxju) ([raw file](https://github.com/robotlocomotion/drake/blob/00cc27b59280badb1dba8c0a0e1429d804a75bb3/drake/geometry/geometry_state.h#L355)):*
> ```objc
>   /** A const iterator through the keys of an unordered map. */
>   template <typename K, typename V>
>   class MapKeyIterator {
> ```

I think the name `MapKeyIterator` is confusing, because this class is not really an iterator. For example, it doesn't provide `operator*`. It is a special wrapper for `std::unordered_map` which uses the custom iterator that you provided. Could you find a better name for the class?

---

*[drake/geometry/geometry_state.h, line 355 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-Ku2UqdksZZOtjsEUPcI:-Ku2UqdksZZOtjsEUPcJ:b-yeutmz) ([raw file](https://github.com/robotlocomotion/drake/blob/00cc27b59280badb1dba8c0a0e1429d804a75bb3/drake/geometry/geometry_state.h#L355)):*
> ```objc
>   /** A const iterator through the keys of an unordered map. */
>   template <typename K, typename V>
>   class MapKeyIterator {
> ```

Can we specify copy/move operations explicitly (i.e. using the macros defined in `drake_copyable.h`) as recommended in Google's C++ guide (for both of `MapKeyIterator` and `iterator`)?

---

*[drake/geometry/geometry_state.h, line 357 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-Ku2QF-NbNpDZbbl0zfr:-Ku2QF-NbNpDZbbl0zfs:b-8sh0oc) ([raw file](https://github.com/robotlocomotion/drake/blob/00cc27b59280badb1dba8c0a0e1429d804a75bb3/drake/geometry/geometry_state.h#L357)):*
> ```objc
>   class MapKeyIterator {
>    public:
>     class iterator {
> ```

BTW, `iterator` -> `Iterator` following our naming-scheme. 
Actually, shouldn't it be called `ConstIterator` because it returns `const K&`, not `K&`?

---

*[drake/geometry/geometry_state.h, line 366 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-Ku2RXH7UbXX4lSodTal:-Ku2RXH7UbXX4lSodTam:bk5vorv) ([raw file](https://github.com/robotlocomotion/drake/blob/00cc27b59280badb1dba8c0a0e1429d804a75bb3/drake/geometry/geometry_state.h#L366)):*
> ```objc
>       bool operator!=(const iterator& other) { return itr_ != other.itr_; }
> 
>      protected:
> ```

I'm wondering why we need `protected:` here. My guess is that `private:` should work as well?

---

*[drake/geometry/geometry_state.h, line 376 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-Ku2Rp5aRSQuLnROvsBO:-Ku2Rp5aRSQuLnROvsBP:b6zj7g4) ([raw file](https://github.com/robotlocomotion/drake/blob/00cc27b59280badb1dba8c0a0e1429d804a75bb3/drake/geometry/geometry_state.h#L376)):*
> ```objc
> 
>     explicit MapKeyIterator(const std::unordered_map<K, V>* map) : map_(map) {}
>     iterator begin() const { return iterator(map_->begin()); }
> ```

BTW, to be explicit, please consider using `->cbegin()` (ditto, `cend()` for the next line).

---

*[drake/geometry/geometry_state.h, line 380 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-Ku2UVb7rvf5sJN7Cb_K:-Ku2UVb7rvf5sJN7Cb_L:bmlon1c) ([raw file](https://github.com/robotlocomotion/drake/blob/00cc27b59280badb1dba8c0a0e1429d804a75bb3/drake/geometry/geometry_state.h#L380)):*
> ```objc
> 
>    private:
>     const std::unordered_map<K, V>* map_;
> ```

Can we change its type to `const std::unordered_map&`? I think it is essentially the same but we don't need to worry about `nullptr`, which we don't have an actual use case here.

---

*[drake/geometry/geometry_system.cc, line 28 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-Ku2VTNi8Tytr9Gam7uj:-Ku2VTNi8Tytr9Gam7uk:b-gx97sy) ([raw file](https://github.com/robotlocomotion/drake/blob/00cc27b59280badb1dba8c0a0e1429d804a75bb3/drake/geometry/geometry_system.cc#L28)):*
BTW, please have `#undef THROW_IF_CONTEXT_ALLOCATED` after the last use of this macro? This practice is recommended at https://google.github.io/styleguide/cppguide.html#Preprocessor_Macros.

> #define macros right before you use them, and #undef them right after.

---

*[drake/geometry/geometry_system.cc, line 277 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-Ku2X9075AFEo2FausLf:-Ku2X9075AFEo2FausLg:b2ti92v) ([raw file](https://github.com/robotlocomotion/drake/blob/00cc27b59280badb1dba8c0a0e1429d804a75bb3/drake/geometry/geometry_system.cc#L277)):*
> ```cpp
>   initial_state_ = nullptr;
>   DRAKE_ASSERT(geometry_state_index_ >= 0);
>   return std::unique_ptr<LeafContext<T>>(
> ```

FYI, the following works as well:
```c++
  return std::make_unique<GeometryContext<T>>(geometry_state_index_);
```
 - It's less verbose (fits in a line!)
 - It's safer (see https://stackoverflow.com/a/22571331)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-:-Ku2Xv5qeQ8UsN-yKNNg:b-dljtwm)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Reviewed 1 of 11 files at r1.
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

*[drake/geometry/geometry_state.h, line 357 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-Ku2QF-NbNpDZbbl0zfr:-Ku2a29TDH9RGtD3WHOD:b-9frf3z) ([raw file](https://github.com/robotlocomotion/drake/blob/00cc27b59280badb1dba8c0a0e1429d804a75bb3/drake/geometry/geometry_state.h#L357)):*
> Actually, shouldn't it be called ConstIterator because it returns const K&, not K&?

I think we need both of `Iterator` and `ConstIterator` classes. The current `iterator` can be renamed to `ConstIterator` while we can add another `Iterator` class. Then we can change the `begin()` of `MapKeyIterator` class to (ditto for `end()`):

```c++
Iterator begin() { return Iterator(map_->begin()); }
ConstIterator begin() const { return ConstIterator(map_->cbegin()); }
```

One motivation for this suggestion is to handle the following code:

```c++
for (FrameId& id : geometry_state_.get_frame_ids()) {
    ...
}
```

which currently generates the following compilation-error:

```
error: binding value of type 'const Identifier<...>' to reference to type 'Identifier<...>' drops 'const' qualifier
``` 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6953)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Reviewed 1 of 6 files at r2.
Review status: all files reviewed at latest revision, 19 unresolved discussions.

---

*[drake/geometry/geometry_visualization.cc, line 23 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-KuJq_AlIo3RKzFa7WJd:-KuJq_AlIo3RKzFa7WJe:by74uma) ([raw file](https://github.com/robotlocomotion/drake/blob/00cc27b59280badb1dba8c0a0e1429d804a75bb3/drake/geometry/geometry_visualization.cc#L23)):*
> ```cpp
> class ShapeToLcm : public ShapeReifier {
>  public:
>   ShapeToLcm() {}
> ```

BTW, we can have ` = default`.

---

*[drake/geometry/geometry_visualization.cc, line 23 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-KuJqfyQ7-sDhKSbkNh8:-KuJqfyQ7-sDhKSbkNh9:b-9mdzyk) ([raw file](https://github.com/robotlocomotion/drake/blob/00cc27b59280badb1dba8c0a0e1429d804a75bb3/drake/geometry/geometry_visualization.cc#L23)):*
> ```cpp
> class ShapeToLcm : public ShapeReifier {
>  public:
>   ShapeToLcm() {}
> ```

FYI, we need explicit copy/move controls here, i.e. `  DRAKE_DEFAULT_COPY_AND_MOVE_AND_ASSIGN(...)`.

---

*[drake/geometry/geometry_visualization.cc, line 23 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-KuJqq-txVyeJvqGMC0G:-KuJqq-txVyeJvqGMC0H:b-4i24e0) ([raw file](https://github.com/robotlocomotion/drake/blob/00cc27b59280badb1dba8c0a0e1429d804a75bb3/drake/geometry/geometry_visualization.cc#L23)):*
> ```cpp
> class ShapeToLcm : public ShapeReifier {
>  public:
>   ShapeToLcm() {}
> ```

FYI, since `ShapeReifier` has a `virtual` destructor. It's recommended to have `~ShapeToLcm() override = default;` here.

---

*[drake/geometry/geometry_visualization.cc, line 47 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-KuJsdqR3udwMsNixVCJ:-KuJsdqR3udwMsNixVCK:bj6kpj3) ([raw file](https://github.com/robotlocomotion/drake/blob/00cc27b59280badb1dba8c0a0e1429d804a75bb3/drake/geometry/geometry_visualization.cc#L47)):*
> ```cpp
> 
>     Eigen::Map<Eigen::Vector4f> color(geometry_data_.color);
>     color = in_color.template cast<float>();
> ```

BTW, nit, no need for `template` here.

---

*[drake/geometry/geometry_visualization.cc, line 121 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-KuJtzZNHtKZC3hFL8x4:-KuJtzZNHtKZC3hFL8x5:bpvnkit) ([raw file](https://github.com/robotlocomotion/drake/blob/00cc27b59280badb1dba8c0a0e1429d804a75bb3/drake/geometry/geometry_visualization.cc#L121)):*
> ```cpp
>       message.link[0].geom.resize(anchored_count);
>       int geom_index = 0;
>       for (const auto &pair : state.anchored_geometries_) {
> ```

BTW, `auto &` -> `auto& `

---

*[drake/geometry/geometry_visualization.cc, line 123 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-KuJu1rs5ypbdVR1fkYm:-KuJu1rs5ypbdVR1fkYn:b-utwuj3) ([raw file](https://github.com/robotlocomotion/drake/blob/00cc27b59280badb1dba8c0a0e1429d804a75bb3/drake/geometry/geometry_visualization.cc#L123)):*
> ```cpp
>       for (const auto &pair : state.anchored_geometries_) {
>         const InternalAnchoredGeometry& geometry = pair.second;
>         const Shape &shape = geometry.get_shape();
> ```

BTW, `Shape &` -> `Shape& `

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6953)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Good eye, @soonho-tri. I think I've hit them all. The one point to quibble is on the `MapKeyIterator` PTAL

---

Review status: 5 of 11 files reviewed at latest revision, 19 unresolved discussions.

---

*[drake/geometry/geometry_context.h, line 19 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-Ku26ZQSNeGKDq1RP9Iw:-KuL-hxtrTIYT2rVqORL:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/00cc27b59280badb1dba8c0a0e1429d804a75bb3/drake/geometry/geometry_context.h#L19)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

Can it be marked as `final`?
</blockquote></details>

Done

---

*[drake/geometry/geometry_context.h, line 19 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-Ku26bQgjTpcpqRYgCbU:-KuL-icYcbDFzyYna3B5:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/00cc27b59280badb1dba8c0a0e1429d804a75bb3/drake/geometry/geometry_context.h#L19)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

FYI, `drake::` is redundant since we're in the namespace.
</blockquote></details>

Done

---

*[drake/geometry/geometry_context.h, line 20 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-Ku26o3My0lZStzsPgqA:-KuL-nSI3VZwspfvgR6y:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/00cc27b59280badb1dba8c0a0e1429d804a75bb3/drake/geometry/geometry_context.h#L20)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

Can we specify copy/move operations explicitly (i.e. using the macros defined in `drake_copyable.h`) as recommended in Google's C++ guide?
</blockquote></details>

Done

---

*[drake/geometry/geometry_context.h, line 32 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-Ku25wQfLVwASNsaDCwB:-KuL-xw4LHYywZWkwuyd:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/00cc27b59280badb1dba8c0a0e1429d804a75bb3/drake/geometry/geometry_context.h#L32)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

FYI, POD member fields need `{}`. 
BTW, this member field can be marked as `const`.
</blockquote></details>

Done

---

*[drake/geometry/geometry_state.h, line 111 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-Ku2Nd2iOFPv4lymMSVJ:-KuL75iZVPtOA_c_HIV9:b3sty49) ([raw file](https://github.com/robotlocomotion/drake/blob/00cc27b59280badb1dba8c0a0e1429d804a75bb3/drake/geometry/geometry_state.h#L111)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

Returning `const T&` can be efficient but it can cause a problem if one uses a reference after the referenced frame was removed. Could you share your thought regarding this concern?
</blockquote></details>

Done This pointed to a larger issue. So, I addressed it here and its fellow methods.

---

*[drake/geometry/geometry_state.h, line 355 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-Ku2SxcvNoBpLXyhaFpe:-KuLimJqnaRBsHKJg8bV:b-j7ffqu) ([raw file](https://github.com/robotlocomotion/drake/blob/00cc27b59280badb1dba8c0a0e1429d804a75bb3/drake/geometry/geometry_state.h#L355)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

I think the name `MapKeyIterator` is confusing, because this class is not really an iterator. For example, it doesn't provide `operator*`. It is a special wrapper for `std::unordered_map` which uses the custom iterator that you provided. Could you find a better name for the class?
</blockquote></details>

So, it would be reasonably accurate to call this a `MapKeyRangeIterator`. I.e., it can be used in range iterations (but nothing else). How do you feel about that?

---

*[drake/geometry/geometry_state.h, line 355 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-Ku2UqdksZZOtjsEUPcI:-KuLkFzDmXKrQiPmO43o:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/00cc27b59280badb1dba8c0a0e1429d804a75bb3/drake/geometry/geometry_state.h#L355)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

Can we specify copy/move operations explicitly (i.e. using the macros defined in `drake_copyable.h`) as recommended in Google's C++ guide (for both of `MapKeyIterator` and `iterator`)?
</blockquote></details>

Done

---

*[drake/geometry/geometry_state.h, line 357 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-Ku2QF-NbNpDZbbl0zfr:-KuLkhFidI9dAmzfZO8S:b-g2wf1f) ([raw file](https://github.com/robotlocomotion/drake/blob/00cc27b59280badb1dba8c0a0e1429d804a75bb3/drake/geometry/geometry_state.h#L357)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

> Actually, shouldn't it be called ConstIterator because it returns const K&, not K&?

I think we need both of `Iterator` and `ConstIterator` classes. The current `iterator` can be renamed to `ConstIterator` while we can add another `Iterator` class. Then we can change the `begin()` of `MapKeyIterator` class to (ditto for `end()`):

```c++
Iterator begin() { return Iterator(map_->begin()); }
ConstIterator begin() const { return ConstIterator(map_->cbegin()); }
```

One motivation for this suggestion is to handle the following code:

```c++
for (FrameId& id : geometry_state_.get_frame_ids()) {
    ...
}
```

which currently generates the following compilation-error:

```
error: binding value of type 'const Identifier<...>' to reference to type 'Identifier<...>' drops 'const' qualifier
``` 
</blockquote></details>

The documentation for the `Identifier` class would consider that to be a misuse. `Identifier` instances should be treated with the same semantics as `int`s.

---

*[drake/geometry/geometry_state.h, line 366 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-Ku2RXH7UbXX4lSodTal:-KuLkxH8C46vf-QwmFIO:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/00cc27b59280badb1dba8c0a0e1429d804a75bb3/drake/geometry/geometry_state.h#L366)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

I'm wondering why we need `protected:` here. My guess is that `private:` should work as well?
</blockquote></details>

Done

---

*[drake/geometry/geometry_state.h, line 376 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-Ku2Rp5aRSQuLnROvsBO:-KuLl0jiLMl_VtT8OPj2:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/00cc27b59280badb1dba8c0a0e1429d804a75bb3/drake/geometry/geometry_state.h#L376)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, to be explicit, please consider using `->cbegin()` (ditto, `cend()` for the next line).
</blockquote></details>

Done

---

*[drake/geometry/geometry_state.h, line 380 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-Ku2UVb7rvf5sJN7Cb_K:-KuLl6U_AM_JVF8U5mEc:b-l6ql) ([raw file](https://github.com/robotlocomotion/drake/blob/00cc27b59280badb1dba8c0a0e1429d804a75bb3/drake/geometry/geometry_state.h#L380)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

Can we change its type to `const std::unordered_map&`? I think it is essentially the same but we don't need to worry about `nullptr`, which we don't have an actual use case here.
</blockquote></details>

Well, we'd have to worry about it just as much; we'd only be moving the source of the concern. Our standard requires to pass in a `const *` if we're going persist a "reference". So, for a pointer input, we can store a pointer or a reference.
- pointer: we get a error when we try to grab an iterator.
- reference: we get an error when we try to store the null pointer as a reference.

That said, I've left it as a pointer and added a `DRAKE_DEMAND` to capture the nullness of it at construction.

---

*[drake/geometry/geometry_system.cc, line 28 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-Ku2VTNi8Tytr9Gam7uj:-KuLlqdWosLN7_YDjS2D:b-7jb2bm) ([raw file](https://github.com/robotlocomotion/drake/blob/00cc27b59280badb1dba8c0a0e1429d804a75bb3/drake/geometry/geometry_system.cc#L28)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, please have `#undef THROW_IF_CONTEXT_ALLOCATED` after the last use of this macro? This practice is recommended at https://google.github.io/styleguide/cppguide.html#Preprocessor_Macros.

> #define macros right before you use them, and #undef them right after.

</blockquote></details>

Done 

(FTR Independent of style, am I mistaken in believing that this would only persist for the actual compilation of `geometry_system.cc` into an object file? Although macros have global scope, the binning of compilation would preclude this macro cluttering up other compilation, yes?)

---

*[drake/geometry/geometry_system.cc, line 277 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-Ku2X9075AFEo2FausLf:-KuLuksuQRdWGQEQEX3F:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/00cc27b59280badb1dba8c0a0e1429d804a75bb3/drake/geometry/geometry_system.cc#L277)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

FYI, the following works as well:
```c++
  return std::make_unique<GeometryContext<T>>(geometry_state_index_);
```
 - It's less verbose (fits in a line!)
 - It's safer (see https://stackoverflow.com/a/22571331)
</blockquote></details>

Done

---

*[drake/geometry/geometry_visualization.cc, line 23 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-KuJq_AlIo3RKzFa7WJd:-KuLux_2-hFH1dN9FM5I:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/00cc27b59280badb1dba8c0a0e1429d804a75bb3/drake/geometry/geometry_visualization.cc#L23)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, we can have ` = default`.
</blockquote></details>

Done

---

*[drake/geometry/geometry_visualization.cc, line 23 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-KuJqfyQ7-sDhKSbkNh8:-KuLuyCl93lxLTvvbwe9:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/00cc27b59280badb1dba8c0a0e1429d804a75bb3/drake/geometry/geometry_visualization.cc#L23)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

FYI, we need explicit copy/move controls here, i.e. `  DRAKE_DEFAULT_COPY_AND_MOVE_AND_ASSIGN(...)`.
</blockquote></details>

Done

---

*[drake/geometry/geometry_visualization.cc, line 23 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-KuJqq-txVyeJvqGMC0G:-KuLv1kH0VTuAsWpkpGo:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/00cc27b59280badb1dba8c0a0e1429d804a75bb3/drake/geometry/geometry_visualization.cc#L23)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

FYI, since `ShapeReifier` has a `virtual` destructor. It's recommended to have `~ShapeToLcm() override = default;` here.
</blockquote></details>

Done

---

*[drake/geometry/geometry_visualization.cc, line 47 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-KuJsdqR3udwMsNixVCJ:-KuLvATpYTv2mvVIzGyI:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/00cc27b59280badb1dba8c0a0e1429d804a75bb3/drake/geometry/geometry_visualization.cc#L47)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, nit, no need for `template` here.
</blockquote></details>

Done

---

*[drake/geometry/geometry_visualization.cc, line 121 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-KuJtzZNHtKZC3hFL8x4:-KuLvEtM0tN2tqhdcpfo:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/00cc27b59280badb1dba8c0a0e1429d804a75bb3/drake/geometry/geometry_visualization.cc#L121)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, `auto &` -> `auto& `
</blockquote></details>

Done

---

*[drake/geometry/geometry_visualization.cc, line 123 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-KuJu1rs5ypbdVR1fkYm:-KuLvJ9-46EtEOi0k3Rn:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/00cc27b59280badb1dba8c0a0e1429d804a75bb3/drake/geometry/geometry_visualization.cc#L123)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, `Shape &` -> `Shape& `
</blockquote></details>

Done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-:-KuLvK_a3Iad7pluDYh4:b-qrqzux)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Reviewed 2 of 6 files at r3.
Review status: 7 of 11 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/geometry/geometry_state.h, line 111 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-Ku2Nd2iOFPv4lymMSVJ:-KuNaJ4-H_IXf9XsOLq_:b-reriv9) ([raw file](https://github.com/robotlocomotion/drake/blob/00cc27b59280badb1dba8c0a0e1429d804a75bb3/drake/geometry/geometry_state.h#L111)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

Done This pointed to a larger issue. So, I addressed it here and its fellow methods.
</blockquote></details>

Nice!
BTW, could you check if doxygen handles `Throws` as `@throws`?

---

*[drake/geometry/geometry_system.cc, line 28 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-Ku2VTNi8Tytr9Gam7uj:-KuNYWtUABZ8zMnTLtAe:b-xlq7tk) ([raw file](https://github.com/robotlocomotion/drake/blob/00cc27b59280badb1dba8c0a0e1429d804a75bb3/drake/geometry/geometry_system.cc#L28)):*
> (FTR Independent of style, am I mistaken in believing that this would only persist for the actual compilation of geometry_system.cc into an object file? Although macros have global scope, the binning of compilation would preclude this macro cluttering up other compilation, yes?)

You're right. Unless one includes this `.cc` file (which is very very unlikely), this macro is visible only in this file, which could be good enough. At the same time, we know that it's always better to keep a scope as small as possible (i.e. global variable vs. local variable) to avoid any (future) problems. I think that's why GCG recommends this. Note that they don't even allow to use macros in a header file.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6953)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Review status: 7 of 11 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/geometry/geometry_state.h, line 111 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-Ku2Nd2iOFPv4lymMSVJ:-KuPU4O5hN54My3SMmKy:b-ys9ybl) ([raw file](https://github.com/robotlocomotion/drake/blob/00cc27b59280badb1dba8c0a0e1429d804a75bb3/drake/geometry/geometry_state.h#L111)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

Nice!
BTW, could you check if doxygen handles `Throws` as `@throws`?
</blockquote></details>

Done I assume it doesn't and decided to make this doxygen look more like its peers.

---

*[drake/geometry/geometry_system.cc, line 28 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-Ku2VTNi8Tytr9Gam7uj:-KuPUGskRL1efbuHfJTi:bvyy8n) ([raw file](https://github.com/robotlocomotion/drake/blob/00cc27b59280badb1dba8c0a0e1429d804a75bb3/drake/geometry/geometry_system.cc#L28)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

> (FTR Independent of style, am I mistaken in believing that this would only persist for the actual compilation of geometry_system.cc into an object file? Although macros have global scope, the binning of compilation would preclude this macro cluttering up other compilation, yes?)

You're right. Unless one includes this `.cc` file (which is very very unlikely), this macro is visible only in this file, which could be good enough. At the same time, we know that it's always better to keep a scope as small as possible (i.e. global variable vs. local variable) to avoid any (future) problems. I think that's why GCG recommends this. Note that they don't even allow to use macros in a header file.
</blockquote></details>

Well, *technically* they say you can have them if you have a note from a technical lead.  :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6953)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Thanks for patience!

---

Reviewed 1 of 6 files at r2, 3 of 6 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/geometry/geometry_state.h, line 355 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-Ku2SxcvNoBpLXyhaFpe:-KuQ4wlS12fC7lTIbPkf:by0ruq7) ([raw file](https://github.com/robotlocomotion/drake/blob/00cc27b59280badb1dba8c0a0e1429d804a75bb3/drake/geometry/geometry_state.h#L355)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

So, it would be reasonably accurate to call this a `MapKeyRangeIterator`. I.e., it can be used in range iterations (but nothing else). How do you feel about that?
</blockquote></details>

I think it's clearer if we avoid the term `Iterator`. How about `MapKeyRangeWrapper` (or a variant of it)?

---

*[drake/geometry/geometry_system.cc, line 28 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-Ku2VTNi8Tytr9Gam7uj:-KuQ5kkdgZKAPYnZoZC3:b-ygj14u) ([raw file](https://github.com/robotlocomotion/drake/blob/00cc27b59280badb1dba8c0a0e1429d804a75bb3/drake/geometry/geometry_system.cc#L28)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

Well, *technically* they say you can have them if you have a note from a technical lead.  :)
</blockquote></details>

:-)

![IMG\_4059.JPG](https://files.reviewableusercontent.io/images/reviewable/20779767/hmAnV9AWE5DLEXGztDATn46F/IMG_4059.jpeg)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-:-KuQ5vSEHJuhbOGof36X:bo5bd81)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Review status: 9 of 11 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/geometry/geometry_state.h, line 355 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6953#-Ku2SxcvNoBpLXyhaFpe:-KuQ6uTpOYBiFFe03zkS:b-sc4gy2) ([raw file](https://github.com/robotlocomotion/drake/blob/00cc27b59280badb1dba8c0a0e1429d804a75bb3/drake/geometry/geometry_state.h#L355)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

I think it's clearer if we avoid the term `Iterator`. How about `MapKeyRangeWrapper` (or a variant of it)?
</blockquote></details>

Given the language here:

http://en.cppreference.com/w/cpp/language/range-for

I think I'd advocate `MapKeyRange`. I've also updated other references, documentation, and names.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6953)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Reviewed 2 of 2 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6953)*
<!-- Sent from Reviewable.io -->

avalenzu commented:
Reviewed 4 of 6 files at r3, 2 of 2 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6953)*
<!-- Sent from Reviewable.io -->
